File: go/pkg/linked/linked_8.go
package linked

// FindLoopStart is O(n)
func FindLoopStart(l List) *Node {
	m := map[*Node]bool{}
	node := l.Head

	for node != nil {
		if m[node] {
			return node
		}
		m[node] = true
		node = node.Next
	}

	return nil
}

// FindLoopStartHard is O(n)
// see comments in Python solution
func FindLoopStartHard(l List) *Node {
	fast := l.Head
	slow := l.Head

	fast = fast.Next.Next
	slow = slow.Next

	for fast != slow {
		fast = fast.Next.Next
		slow = slow.Next
	}

	slow = l.Head

	for fast != slow {
		fast = fast.Next
		slow = slow.Next
	}

	return fast
}

File: ts/solutions/linked/linked_8.ts
import { LinkedList, Node_ } from './linked_lists_1'

// findLoopStart is O(n)
function findLoopStart(l: LinkedList): Node_ {
  const visited: Set<Node_> = new Set()
  let node = l.head

  while (node) {
    if (visited.has(node)) {
      return node
    }
    visited.add(node)
    node = node.next
  }
}

// findLoopStartHard is O(n)
// see comments in Python solution
function findLoopStartHard(l: LinkedList): Node_ {
  let fast = l.head
  let slow = l.head

  fast = fast.next.next
  slow = slow.next

  while (fast !== slow) {
    fast = fast.next.next
    slow = slow.next
  }

  slow = l.head

  while (fast !== slow) {
    fast = fast.next
    slow = slow.next
  }

  return fast
}
